core/src/main/java/bisq/core/dao/state/unconfirmed/UnconfirmedBsqChangeOutputListService.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * This file is part of Bisq.
+ *
+ * Bisq is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or (at
+ * your option) any later version.
+ *
+ * Bisq is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with Bisq. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package bisq.core.dao.state.unconfirmed;
+
+import bisq.core.dao.state.model.blockchain.TxType;
+
+import bisq.common.app.DevEnv;
+import bisq.common.proto.persistable.PersistedDataHost;
+import bisq.common.storage.Storage;
+
+import org.bitcoinj.core.Coin;
+import org.bitcoinj.core.Transaction;
+import org.bitcoinj.core.TransactionConfidence;
+import org.bitcoinj.core.TransactionInput;
+import org.bitcoinj.core.TransactionOutput;
+import org.bitcoinj.wallet.Wallet;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Objects;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class UnconfirmedBsqChangeOutputListService implements PersistedDataHost {
+    private UnconfirmedBsqChangeOutputList unconfirmedBsqChangeOutputList = new UnconfirmedBsqChangeOutputList();
+    private final Storage<UnconfirmedBsqChangeOutputList> storage;
+
+    @Inject
+    public UnconfirmedBsqChangeOutputListService(Storage<UnconfirmedBsqChangeOutputList> storage) {
+        this.storage = storage;
+    }
+
+    @Override
+    public void readPersisted() {
+        if (DevEnv.isDaoActivated()) {
+            UnconfirmedBsqChangeOutputList persisted = storage.initAndGetPersisted(unconfirmedBsqChangeOutputList, 100);
+            if (persisted != null) {
+                unconfirmedBsqChangeOutputList.clear();
+                unconfirmedBsqChangeOutputList.addAll(persisted.getList());
+            }
+        }
+    }
+
+
+    ///////////////////////////////////////////////////////////////////////////////////////////
+    // API
+    ///////////////////////////////////////////////////////////////////////////////////////////
+
+    /**
+     * Once a tx gets committed to out BSQ wallet we store the change output for allowing it to be spent in follow-up
+     * transactions.
+     */
+    public void onCommitTx(Transaction tx, TxType txType, Wallet wallet) {
+        // We remove all potential connected outputs from our inputs as they would have been spent.
+        removeConnectedOutputsOfInputsOfTx(tx);
```

Comment:
Ok, I see the outputs are removed when spent, previous comment is then void since all transactions in the unconfirmed list are unspent.

Reply:
Yes, if data hanlding stays in sync it should be ok. Still as we manage parallel data now there might be theoretically an issue (e.g. if persistence fails after remove). I will try to test that scenario to verify that we get an exception from BitcoinJ.